data_pipeline_api/registry/downloader.py

```diff
@@ -0,0 +1,372 @@
+        self._token: str = token or os.environ.get(DATA_REGISTRY_ACCESS_TOKEN)
+
+        self._data_products: List[Dict[Tuple[str, str], str]] = []
+        self._external_objects: List[Dict[Tuple[str, str], str]] = []
+        self._resolved_data_products: List[Dict[Tuple[str, str], Any]] = []
+        self._resolved_external_objects: List[Dict[Tuple[str, str], Any]] = []
+
+    def add_data_product(
+        self, namespace: str, data_product: str, component: Optional[str] = None, version: Optional[str] = None
+    ) -> None:
+        """
+        Registers a data product to be downloaded
+        """
+        self._data_products.append(
+            {
+                (DataRegistryTarget.namespace, DataRegistryField.name): namespace,
+                (DataRegistryTarget.data_product, DataRegistryField.name): data_product,
+                (DataRegistryTarget.object_component, DataRegistryField.name): component,
+                (DataRegistryTarget.data_product, DataRegistryField.version): version,
+            }
+        )
+
+    def add_external_object(self, doi_or_unique_name: str, version: Optional[str] = None) -> None:
+        """
+        Registers an external object to be downloaded
+        """
+        self._external_objects.append(
+            {
+                (DataRegistryTarget.external_object, DataRegistryField.doi_or_unique_name): doi_or_unique_name,
+                (DataRegistryTarget.external_object, DataRegistryField.version): version,
+            }
+        )
+
+    def _resolve_namespaces(self, input_blocks: List[Dict[Tuple[str, str], Any]]) -> List[Dict[Tuple[str, str], Any]]:
+        resolved = []
+        for block in input_blocks:
+            namespace_name = block[DataRegistryTarget.namespace, DataRegistryField.name]
+            namespaces = get_data(
+                {DataRegistryField.name: namespace_name},
+                DataRegistryTarget.namespace,
+                self._data_registry_url,
+                self._token,
+                exact=False,
+            )
+            if namespaces:
+                for namespace in namespaces:
+                    cblock = block.copy()
+                    for k, v in namespace.items():
+                        cblock[DataRegistryTarget.namespace, k] = v
+                    resolved.append(cblock)
+        return resolved
+
+    def _resolve_data_products(
+        self, input_blocks: List[Dict[Tuple[str, str], Any]]
+    ) -> List[Dict[Tuple[str, str], Any]]:
+        resolved = []
+        for block in input_blocks:
+            query_data = {
+                DataRegistryField.name: block[DataRegistryTarget.data_product, DataRegistryField.name],
+                DataRegistryField.namespace: block[DataRegistryTarget.namespace, DataRegistryField.url],
+            }
+            version = block[DataRegistryTarget.data_product, DataRegistryField.version]
```

Comment:
do you want this to be a `get`? (asking because of the line below)

Reply:
there are a few similar cases in the code, so I'm going to assume the answer is no, but probably worth confirming